docs/examples/custom_channel.py

```diff
@@ -0,0 +1,90 @@
+#
+# Copyright (c) 2024 ZettaScale Technology
+#
+# This program and the accompanying materials are made available under the
+# terms of the Eclipse Public License 2.0 which is available at
+# http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+# which is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+#
+# Contributors:
+#   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+#
+import queue
+import threading
+import time
+from collections.abc import Callable
+from typing import Generic, TypeVar, Union
+
+import zenoh
+
+_T = TypeVar("_T")
+
+
+# Test support: send data in background
+def send_data():
+    time.sleep(3)
+    for i in range(2):
+        session.put("key/expression", f"sample_{i}")
+
+
+threading.Thread(target=send_data, daemon=True).start()
+
+
+# [custom_channel]
+class PriorityChannel(Generic[_T]):
+    def __init__(self, maxsize=100):
+        self.queue: queue.PriorityQueue = queue.PriorityQueue(maxsize)
+        # Counter to preserve FIFO order for samples with same priority
+        self._counter = 0
+
+    def recv(self) -> _T:
+        return self.queue.get()[2]
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> _T:
+        sample = self.recv()
+        if sample is None:
+            raise StopIteration
+        return sample
```

Comment:
I don't think this part is necessary for the sake of the example. We should just demonstrate a minimal working example. We don't even know if this use case is even real. Let's wait for people to ask for this kind of refinement if they need it.